src/main/kotlin/mcts/MoteCarloTreeSearch.kt

```diff
@@ -0,0 +1,87 @@
+package mcts
+
+import mcts.util.Duration
+import mcts.util.Random
+
+object MonteCarloTreeSearch {
+    private val WIN_SCORE = 10.0
+
+    fun findNextMove(board: Board, player: Player, random: Random, duration: Duration): Board {
+        val start = System.currentTimeMillis()
+        val end = start + duration.toMillis()
+
+        val opponent = player.opponent
+        val rootNode = Node(State(board, opponent))
+
+        while (System.currentTimeMillis() < end) {
+            // Phase 1 - Selection
+            val promisingNode = selectPromisingNode(rootNode)
+            // Phase 2 - Expansion
+            val status = promisingNode.state.board.checkStatus()
+            if (status is Status.InProgress)
+                expandNode(promisingNode, status.positions)
+
+            // Phase 3 - Simulation
+            val nodeToExplore = if (promisingNode.children.isNotEmpty()) {
+                promisingNode.randomChildNode(random)
+            } else promisingNode
+
+            val playoutResult = simulateRandomPlayout(nodeToExplore, opponent, random)
+            // Phase 4 - Update
+            backPropogation(nodeToExplore, playoutResult)
+        }
+
+        val winnerNode = rootNode.childWithMaxScore
+        return winnerNode.state.board
+    }
+
+    private fun selectPromisingNode(rootNode: Node): Node {
+        var node = rootNode
+        while (node.children.isNotEmpty()) {
+            node = UCT.findBestNodeWithUCT(node)
+        }
+        return node
+    }
+
+    private fun expandNode(node: Node, positions: List<Position>) {
+        val possibleStates = node.state.allPossibleStates(positions)
+        possibleStates.forEach { state ->
```

Comment:
Her lager du nye states som tydeligvis ikke er gyldige, for du må bytte spiller etterpå. Tipper `node.state.player.opponent` burde være et parameter til `allPossibleStates`